### PR TITLE
Add feature tests via Behat

### LINKED ISSUE #3076: Add functional tests for CLI commands
The obvious choice would be reusing `wp-cli/wp-cli-tests` to write Behat tests, but there seems to be a blocking technical issue so far that doesn't allow this without resorting to a hack.

The main issue is that you need to install the plugin within the WordPress installation that Behat will set up. To do so, you use a built plugin version and send it over for installation as a ZIP archive.

However, this built version will contain the same Composer autoloader (with the same hash included in the class name) than the Composer autoloader that Behat was actually run from. This causes a redeclaration error from PHP when the WordPress installation tries to load the plugin.

I can think of two possible solutions right now to solve this:
1.) Rebuild a fresh ZIP archive with a regenerated autoloader that will have a different hash than the one Behat was loaded from.
2.) Avoid using Behat directly as a Composer dependency and use a binary/Phar instead, that comes with its own autoloader.

Before trying to tackle this, I'd like to investigate this from the WP-CLI side, to see whether this can be streamlined through the official package first.

Related #1815

### PR DESCRIPTION
**Pull request by schlessera**
## Summary

This PR adds feature testing via Behat to the plugin. It reuses the testing infrastructure and Behat definitions from WP-CLI (see https://github.com/wp-cli/wp-cli-tests).

Before the tests are run for the first time, the following command has to be executed:
```bash
composer prepare-tests
```

This should create the needed table(s) in the database for the tests to be run within.

Database credentials can be provided via environment variables if the database is not found on `localhost:3306`:
- `WP_CLI_TEST_DBHOST` is the host to use and can include a port, i.e `127.0.0.1:33060` (defaults to `localhost`)
- `WP_CLI_TEST_DBROOTUSER` is the user that has permission to administer databases and users (defaults to `root`).
- `WP_CLI_TEST_DBROOTPASS` is the password to use for the above user (defaults to an empty password).
- `WP_CLI_TEST_DBUSER` is the user that the tests run under (defaults to `wp_cli_test`).
- `WP_CLI_TEST_DBPASS` is the password to use for the above user (defaults to `password1`).

Then, to run the tests, the following command can be used:
```bash
vendor/bin/behat
```

**Note:** The code coverage is currently not being executed on GHA because there is still an issue with the interpretation of the code coverage data within codecov (probably a path discrepancy). This should be fixed and re-enabled in a separate PR.

Fixes #3076

## Checklist

- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

### CODE REVIEW

File: includes/bootstrap.php [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+	__DIR__ . '/../includes/amp-helper-functions.php' => 'amp_activate',
+	__DIR__ . '/../includes/admin/functions.php'      => 'amp_init_customizer',
+	__DIR__ . '/../includes/deprecated.php'           => 'amp_load_classes',
+];
+
+foreach ( $files_to_include as $file_to_include => $function_to_check ) {
+	if ( ! function_exists( $function_to_check ) ) {
+		include $file_to_include;
+	}
+}
```
Comment (westonruter) [inline includes/bootstrap.php:23]: Wouldn't this be simpler as:

```suggestion
$files_to_include = [
	__DIR__ . '/../back-compat/back-compat.php',
	__DIR__ . '/../includes/amp-helper-functions.php',
	__DIR__ . '/../includes/admin/functions.php',
	__DIR__ . '/../includes/deprecated.php',
];

foreach ( $files_to_include as $file_to_include ) {
	require_once $file_to_include;
}
```

Comment (westonruter) [reply includes/bootstrap.php:23]: Or even simpler:

```suggestion
require_once __DIR__ . '/../back-compat/back-compat.php';
require_once __DIR__ . '/../includes/amp-helper-functions.php';
require_once __DIR__ . '/../includes/admin/functions.php';
require_once  __DIR__ . '/../includes/deprecated.php';
```

File: tests/php/maybe-generate-wp-cli-coverage.php [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+use SebastianBergmann\CodeCoverage\Filter;
+use SebastianBergmann\CodeCoverage\Driver\Selector;
+use SebastianBergmann\CodeCoverage\CodeCoverage;
+use SebastianBergmann\CodeCoverage\Report\Clover;
+
+$root_folder = realpath( dirname( dirname( __DIR__ ) ) );
+
+if ( ! class_exists( 'SebastianBergmann\CodeCoverage\Filter' ) ) {
+	require "{$root_folder}/vendor/autoload.php";
```
Comment (westonruter) [inline tests/php/maybe-generate-wp-cli-coverage.php:11]:
```suggestion
	require_once "{$root_folder}/vendor/autoload.php";
```

**Comment by westonruter** [comment]
I tried running `composer prepare-tests` but I got:

```
> install-package-tests
ERROR 2002 (HY000): Can't connect to local MySQL server through socket '/var/run/mysqld/mysqld.sock' (2)
Detected MySQL at version .
/app/public/content/plugins/amp/vendor/bin/install-package-tests: 75: [: Illegal number:
+ mysql -e CREATE DATABASE IF NOT EXISTS `wp_cli_test`; -uroot
ERROR 2002 (HY000): Can't connect to local MySQL server through socket '/var/run/mysqld/mysqld.sock' (2)
Script install-package-tests handling the prepare-tests event returned with error code 1
```

**Comment by westonruter** [comment]
I guess that's because my environment variables are empty.

File: tests/php/maybe-generate-wp-cli-coverage.php (hunk 2 above)
Comment (schlessera) [reply tests/php/maybe-generate-wp-cli-coverage.php:11]: As we already check whether the class exists, the `require_once` is only slower than `require` in the case of the expected behavior, and might cloak a problem if the dependency is actually missing. In that case, I'd prefer for the code to fail right away in here to show that class was missing even though the autoloader was already required before.

File: includes/bootstrap.php (hunk 1 above)
Comment (schlessera) [reply includes/bootstrap.php:23]: That could potentially fail in several different scenarios, due to the way the Composer `'files'` directive works.

The only safe behavior is to write the code in such a way that is independent of both loading order and filesystem location. For example, it cannot be assumed that there's only one version of these files, and the `require_once` only checks for the file path, not what it is actually supposed to declare.

Comment (westonruter) [reply includes/bootstrap.php:23]: Got it.

**Review by westonruter — APPROVED** [review]
Tests are working in CI context, so that's the most important thing.

**Pull request merged**